Jenkinsfile

```diff
@@ -25,19 +15,99 @@ pipeline {
     }    
 
     stages {
-        stage('Run SonarQube analysis') {
+        stage('Build and Test Locally') {
+            stages {
+                stage('Build Docker Images') {
+                    steps {
+                        script {
+                            sh label: 'Stopping running containers (preventative maintenance)', script: 'docker-compose down -v'
+                            sh label: 'Running docker-compose build', script: 'docker-compose build --build-arg BUILD_TAG=${BUILD_TAG}'
+                        }
+                    }
+                }
+                stage('Deploy Locally') {
+                    steps {
+                        sh label: 'Starting containers', script: 'docker-compose up -d rabbitmq dynamodb nhais'
+                        echo "Waiting 10 seconds for containers to start"
+                        sleep 10
+                        sh label: 'Show all running containers', script: 'docker ps'
+                    }
+                }
+                stage('Run tests') {
+                    steps {
+                        sh label: 'Running tests', script: 'docker-compose run nhais-tests'
+                    }
+                    post {
+                        always {
+                            sh label: 'Copy test reports to folder', script: 'docker cp "$(docker ps -lq)":/usr/src/app/nhais/test-reports .'
+                            junit '**/test-reports/*.xml'
+                            sh label: 'Copy test coverage to folder', script: 'docker cp "$(docker ps -lq)":/usr/src/app/nhais/coverage.xml ./coverage.xml'
+                            cobertura coberturaReportFile: '**/coverage.xml'
+                        }
+                    }
+                }
+                stage('Push image') {
+                    when {
+                        expression { currentBuild.resultIsBetterOrEqualTo('SUCCESS') }
```

Comment:
Unless we do not use parallel, or we do not specify it explicitly the run will stop at the stage that failed and will not progress further, so this may not be needed.

Reply:
Not the case for unstable builds. This is so we don't continue to push images or deploy if there are failing tests.

Reply:
We had a test run for a unit test failing and without this line it continued on with the other stages. Thoughts?

Reply:
May be due to the post { always at the test stage, or the shell command not properly returning status.